### DISCUSSION TITLE
/openapi.json "HTTP Error 500" when a response_class Union is used.

### DISCUSSION THREAD

**Discussion by tacozMacleo** [fastapi/fastapi#7308] [Questions]
## First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [ ] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui). (Happens before this is relevant)
* [ ] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc). (Happens before this is relevant)
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

Here's a self-contained example with my use case:

```Python
from typing import Union

from fastapi import FastAPI, Request
from fastapi.responses import PlainTextResponse
from fastapi.responses import JSONResponse

app = FastAPI()

@app.get("/items/{item_id}", response_class=Union[PlainTextResponse, JSONResponse])
def read_root(
    item_id: str,
    request: Request
):
    if 'application/json' in request.headers["accept"]:
        return JSONResponse({"item_id": item_id})
    elif "text/plain" in request.headers["accept"]:
        return PlainTextResponse("Plain Text")
```

### Description
When open the browser and call the endpoint `/docs` or `/redoc`, I get on redoc a:
```
Something went wrong...
Error downloading http://127.0.0.1:8000/openapi.json HTTP ERROR 500

Stack trace

s/<@https://cdn.jsdelivr.net/npm/redoc@next/bundles/redoc.standalone.js:44:26651
e/</<@https://cdn.jsdelivr.net/npm/redoc@next/bundles/redoc.standalone.js:95:36080


s/<@https://cdn.jsdelivr.net/npm/redoc@next/bundles/redoc.standalone.js:44:26651
e/</<@https://cdn.jsdelivr.net/npm/redoc@next/bundles/redoc.standalone.js:95:35658

ReDoc Version: 2.0.0-rc.40
Commit: 17b9873
```
And on `/docs` I get a:
```
Failed to load API definition.

Errors

 
Fetch error
Internal Server Error /openapi.json
```

But works as expected when I use curl:
```curl -X GET "http://127.0.0.1:8000/items/sdfghj6ty" -H  "accept: text/plain```
or
```curl -X GET "http://127.0.0.1:8000/items/sdfghj6ty" -H  "accept: application/json"```

If I have a Union with only one element in, the error do not occur.
I have not found anywhere how to have multiple response_classes, but it is in the cards that Union is used for this. (Let me know if I am wrong.)

### Environment
* OS: Linux 5.4.0-47-generic #51-Ubuntu SMP Fri Sep 4 19:50:52 UTC 2020 x86_64 GNU/Linux
* FastAPI Version: 0.61.0
* Python Version:  3.8.2

### Additional context
```
$ uvicorn test-request:app --debug
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
INFO:     Started reloader process [490052] using statreload
INFO:     Started server process [490054]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     127.0.0.1:51140 - "GET /docs HTTP/1.1" 200 OK
INFO:     127.0.0.1:51140 - "GET /openapi.json HTTP/1.1" 500 Internal Server Error
ERROR:    Exception in ASGI application
Traceback (most recent call last):
  File "/home/bach/.local/lib/python3.8/site-packages/uvicorn/protocols/http/httptools_impl.py", line 390, in run_asgi
    result = await app(self.scope, self.receive, self.send)
  File "/home/bach/.local/lib/python3.8/site-packages/uvicorn/middleware/proxy_headers.py", line 45, in __call__
    return await self.app(scope, receive, send)
  File "/home/bach/.local/lib/python3.8/site-packages/uvicorn/middleware/debug.py", line 81, in __call__
    raise exc from None
  File "/home/bach/.local/lib/python3.8/site-packages/uvicorn/middleware/debug.py", line 78, in __call__
    await self.app(scope, receive, inner_send)
  File "/home/bach/.local/lib/python3.8/site-packages/fastapi/applications.py", line 179, in __call__
    await super().__call__(scope, receive, send)
  File "/home/bach/.local/lib/python3.8/site-packages/starlette/applications.py", line 111, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/home/bach/.local/lib/python3.8/site-packages/starlette/middleware/errors.py", line 181, in __call__
    raise exc from None
  File "/home/bach/.local/lib/python3.8/site-packages/starlette/middleware/errors.py", line 159, in __call__
    await self.app(scope, receive, _send)
  File "/home/bach/.local/lib/python3.8/site-packages/starlette/exceptions.py", line 82, in __call__
    raise exc from None
  File "/home/bach/.local/lib/python3.8/site-packages/starlette/exceptions.py", line 71, in __call__
    await self.app(scope, receive, sender)
  File "/home/bach/.local/lib/python3.8/site-packages/starlette/routing.py", line 566, in __call__
    await route.handle(scope, receive, send)
  File "/home/bach/.local/lib/python3.8/site-packages/starlette/routing.py", line 227, in handle
    await self.app(scope, receive, send)
  File "/home/bach/.local/lib/python3.8/site-packages/starlette/routing.py", line 41, in app
    response = await func(request)
  File "/home/bach/.local/lib/python3.8/site-packages/fastapi/applications.py", line 128, in openapi
    return JSONResponse(self.openapi())
  File "/home/bach/.local/lib/python3.8/site-packages/fastapi/applications.py", line 106, in openapi
    self.openapi_schema = get_openapi(
  File "/home/bach/.local/lib/python3.8/site-packages/fastapi/openapi/utils.py", line 348, in get_openapi
    result = get_openapi_path(route=route, model_name_map=model_name_map)
  File "/home/bach/.local/lib/python3.8/site-packages/fastapi/openapi/utils.py", line 163, in get_openapi_path
    route_response_media_type: Optional[str] = route.response_class.media_type
  File "/usr/lib/python3.8/typing.py", line 756, in __getattr__
    return getattr(self.__origin__, attr)
AttributeError: '_SpecialForm' object has no attribute 'media_type'
```

**Comment by ArcLightSlavik** [comment]
Looking at the code
```
        response_class: Type[Response] = None,
```
You can only provide a single **response_class**. 
Which would align with OpenAPI (https://github.com/OAI/OpenAPI-Specification/issues/56)
You can use Union to combine multiple classes but this is for **response_model**. 
Example of this: https://github.com/tiangolo/fastapi/pull/108/files.

**Comment by tacozMacleo** [comment]
> Looking at the code
> 
> ```
>         response_class: Type[Response] = None,
> ```
> 
> You can only provide a single **response_class**.
> Which would align with OpenAPI ([OAI/OpenAPI-Specification#56](https://github.com/OAI/OpenAPI-Specification/issues/56))
> You can use Union to combine multiple classes but this is for **response_model**.
> Example of this: https://github.com/tiangolo/fastapi/pull/108/files.

Well it do not say anything about Union in the **response_model**:
```
response_model: Optional[Type[Any]] = None
```
So did not except it to mean anything, that it did not have Union in **response_class**.

**Comment by tacozMacleo** [comment]
Well I quest it have become more of a feature request, to have multiple `response class`, so the Swagger UI would of it.
At least  `response class` are not need to be set, to be able multiple response class types.

**Comment by akshaynarula** [comment]
Something went wrong...
Error downloading http://127.0.0.1:8000/openapi.json HTTP ERROR 500
Stack trace
Error: Error downloading http://127.0.0.1:8000/openapi.json 
HTTP ERROR 500
    at https://cdn.jsdelivr.net/npm/redoc@next/bundles/redoc.standalone.js:29:28049
    at https://cdn.jsdelivr.net/npm/redoc@next/bundles/redoc.standalone.js:80:36080

Error: HTTP ERROR 500
    at https://cdn.jsdelivr.net/npm/redoc@next/bundles/redoc.standalone.js:29:28049
    at https://cdn.jsdelivr.net/npm/redoc@next/bundles/redoc.standalone.js:80:35658

**Comment by tiangolo** [comment]
Thanks for the help here everyone! :clap: :bow: 

Thanks for closing the issue :+1:

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.

**Comment by danvinhe** [comment]
Something went wrong...
I just wrote a basic test case。
Version: **0.128.0**

```
from typing import Union

from fastapi import FastAPI

app = FastAPI()


@app.get("/")
def read_root():
    return {"Hello": "World"}


@app.get("/items/{item_id}")
def read_item(item_id: int, q: Union[str, None] = None):
    return {"item_id": item_id, "q": q}
```
The error is as follows

```
  File "/Users/lindan.he/.pyenv/versions/3.9.0/lib/python3.9/site-packages/pydantic/json_schema.py", line 333, in build_schema_type_to_method
    method_name = f'{key.replace("-", "_")}_schema'
  File "/Users/lindan.he/.pyenv/versions/3.9.0/lib/python3.9/typing.py", line 647, in __getattr__
    return getattr(self.__origin__, attr)
AttributeError: '_SpecialForm' object has no attribute 'replace'
```

**Reply by daverck** [reply]
@danvinhe You need to update your python version
